### PR TITLE
chore: Upgrades serverless_instance resource to auto-generated SDK

### PR DESCRIPTION
**Pull request by lantoli**
## Description

Upgrades serverless_instance resource to auto-generated SDK.

Migration tests created.

Link to any related issue(s): CLOUDP-226092

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/service/serverlessinstance/data_source_serverless_instances.go [hunk 1]
```diff
@@ -25,91 +24,87 @@ func PluralDataSource() *schema.Resource {
...
 
-	list = append(list, serverlessInstances.Results...)
-	obtainedItemsCount += len(serverlessInstances.Results)
+	list = append(list, serverlessInstances.GetResults()...)
+	obtainedItemsCount += len(serverlessInstances.GetResults())
 
-	if serverlessInstances.TotalCount > options.ItemsPerPage && obtainedItemsCount < serverlessInstances.TotalCount {
-		instances, err := getServerlessList(ctx, meta, projectID, options, obtainedItemsCount)
+	if serverlessInstances.GetTotalCount() > *options.ItemsPerPage && obtainedItemsCount < *serverlessInstances.TotalCount {
+		instances, err := getServerlessList(ctx, connV2, options, obtainedItemsCount)
```
Comment (oarbusi) [inline internal/service/serverlessinstance/data_source_serverless_instances.go:76]: in my opinion using recursion here is unnecessary and makes the code more confusing and not straight forward. I would change it to a for loop [equivalent to a "do while"](https://stackoverflow.com/questions/32834661/how-to-replicate-do-while-in-go/32844744#32844744) (so we always do the first iteration of the loop)

File: internal/service/serverlessinstance/resource_serverless_instance.go [hunk 2]
```diff
@@ -165,214 +120,244 @@ func returnServerlessInstanceSchema() map[string]*schema.Schema {
...
+		stateConf := &retry.StateChangeConf{
+			Pending:    []string{"CREATING", "UPDATING", "REPAIRING", "REPEATING", "PENDING"},
+			Target:     []string{"IDLE"},
+			Refresh:    resourceRefreshFunc(ctx, d.Get("name").(string), projectID, connV2),
+			Timeout:    3 * time.Hour,
+			MinTimeout: 1 * time.Minute,
+			Delay:      3 * time.Minute,
+		}
+
+		// Wait, catching any errors
```
Comment (oarbusi) [inline internal/service/serverlessinstance/resource_serverless_instance.go]: is it worth it to keep this comment?

File: internal/service/serverlessinstance/resource_serverless_instance.go [hunk 3]
```diff
@@ -165,214 +120,244 @@ func returnServerlessInstanceSchema() map[string]*schema.Schema {
...
+		Target:     []string{"DELETED"},
+		Refresh:    resourceRefreshFunc(ctx, serverlessName, projectID, connV2),
 		Timeout:    3 * time.Hour,
-		MinTimeout: 1 * time.Minute,
-		Delay:      3 * time.Minute,
+		MinTimeout: 30 * time.Second,
+		Delay:      1 * time.Minute, // Wait 30 secs before starting
 	}
 
 	// Wait, catching any errors
```
Comment (oarbusi) [inline internal/service/serverlessinstance/resource_serverless_instance.go]: same as previous comment

File: internal/service/serverlessinstance/data_source_serverless_instances.go (hunk 1 above)
Comment (lantoli) [reply internal/service/serverlessinstance/data_source_serverless_instances.go:76]: to be honest I just "mimicked" the current code using connv2,  as we'll have a coming project to improve plural data sources. so i'd prefer to leave this optimization for a future PR

File: internal/service/serverlessinstance/resource_serverless_instance.go (hunk 3 above)
Comment (lantoli) [reply internal/service/serverlessinstance/resource_serverless_instance.go]: thx, removed, and the comment is wrong

File: internal/service/serverlessinstance/resource_serverless_instance.go (hunk 2 above)
Comment (lantoli) [reply internal/service/serverlessinstance/resource_serverless_instance.go]: thx, removed

**Review by oarbusi — APPROVED** [review]
LGTM

**Pull request merged**